### PR TITLE
add traceparent header to apache-httpclinet

### PR DESCRIPTION
**Pull request by sekikatsu36**
Same as [RestTemplate interceptor in m3-tracing:spring-web](https://github.com/m3dev/m3-tracing/pull/4), this PR makes m3-tracing:apache-httpclient to support distributed tracing across services by putting `traceparent` header to requests generated by the apache HttpClient.

### CODE REVIEW

File: jvm/apache-httpclient/README.md [hunk 1]
```diff
@@ -4,8 +4,20 @@
...
 httpclient.addRequestInterceptor(M3TracingHttpInterceptor.INSTANCE);
 httpclient.addResponseInterceptor(M3TracingHttpInterceptor.INSTANCE);
 ```
+
+### httpclient 4.3+
+
+```java
+CloseableHttpClient httpClient = HttpClientBuilder.create()
+    .addInterceptorFirst((HttpRequestInterceptor) M3TracingHttpInterceptor.getINSTANCE())
+    .addInterceptorLast((HttpResponseInterceptor) M3TracingHttpInterceptor.getINSTANCE())
```
Comment (saiya) [inline jvm/apache-httpclient/README.md:20]: /`.getINSTANCE()`/`. INSTANCE`/g ?

File: jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    override fun <T> tryGetMetadata(key: HttpRequestMetadataKey<T>): T? = when (key) {
+        HttpRequestMetadataKey.Method -> req.requestLine?.method as T?
+        HttpRequestMetadataKey.Path -> getPath() as T?
+        HttpRequestMetadataKey.Host -> if (context is HttpClientContext) context.targetHost?.hostName as T? else null
+        HttpRequestMetadataKey.Url -> if (context is HttpClientContext) "${context.targetHost?.toURI()}${getPath() ?: ""}" as T? else null
+
+        else -> null
+    }
+
+    // When HttpRequest.requestLine.url is used, remove querystring since the return value of this method may be used by a span name
```
Comment (saiya) [inline jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt:24]: "Path" of HTTP request does not contain parameter. Thus `getPath` method SHOULD remove query parameter always (regardless of the `span name` matter).

File: jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+open class ApacheHttpRequestInfo(protected val req: HttpRequest, protected val context: HttpContext) : HttpRequestInfo {
+    override fun tryGetHeader(name: String): String? = req.getFirstHeader(name)?.value
+    override fun trySetHeader(name: String, value: String) = req.setHeader(name, value)
+
+    @Suppress("UNCHECKED_CAST", "IMPLICIT_ANY")
+    override fun <T> tryGetMetadata(key: HttpRequestMetadataKey<T>): T? = when (key) {
+        HttpRequestMetadataKey.Method -> req.requestLine?.method as T?
+        HttpRequestMetadataKey.Path -> getPath() as T?
+        HttpRequestMetadataKey.Host -> if (context is HttpClientContext) context.targetHost?.hostName as T? else null
+        HttpRequestMetadataKey.Url -> if (context is HttpClientContext) "${context.targetHost?.toURI()}${getPath() ?: ""}" as T? else null
```
Comment (saiya) [inline jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt:19]: (question) Could you tell me why `.toURI()` is needed?

Does `context.targetHost` (== `context.targetHost?.toString()`) returns undesirable string?

File: jvm/apache-httpclient/src/test/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfoTest.kt [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        val uri = mock(URI::class.java)
+        val host = mock(HttpHost::class.java)
+
+        Mockito.`when`(uriRequest.requestLine).thenReturn(requestLine)
+        Mockito.`when`(uriRequest.uri).thenReturn(uri)
+        Mockito.`when`(clientContext.targetHost).thenReturn(host)
+        Mockito.`when`(uri.path).thenReturn("/foo/bar.html")
+        Mockito.`when`(requestLine.method).thenReturn("GET")
+        Mockito.`when`(host.hostName).thenReturn("test.m3.com")
+        Mockito.`when`(host.toURI()).thenReturn("http://test.m3.com")
```
Comment (saiya) [inline jvm/apache-httpclient/src/test/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfoTest.kt:47]: Could you use RFC defined example domain for test purpose?
( "example.com" or it's subdomain )

File: jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt (hunk 2 above)
Comment (sekikatsu36) [reply jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt]: change the comment

File: jvm/apache-httpclient/src/test/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfoTest.kt (hunk 4 above)
Comment (sekikatsu36) [reply jvm/apache-httpclient/src/test/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfoTest.kt]: I have changed the domain to `example.com`

File: jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt (hunk 3 above)
Comment (sekikatsu36) [reply jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt]: I'd not noticed `toString()` is enough, so changed from `toURI` to `toString`

File: jvm/apache-httpclient/README.md (hunk 1 above)
Comment (sekikatsu36) [reply jvm/apache-httpclient/README.md:20]: スクリーンショット 2020-07-31 20 05 49

For now, I need to use `getInstance()` as shown above.

The code decompiled to java is follow:

```java
private static final M3TracingHttpInterceptor INSTANCE = new M3TracingHttpInterceptor();
```

To access `INSTANCE`, the field seems to be annotated `@jvmField` not `@jvmStatic`

kotlin
```kotlin
@JvmField
public val INSTANCE = M3TracingHttpInterceptor()
```

java
```java
public static final M3TracingHttpInterceptor INSTANCE = new M3TracingHttpInterceptor();
```

Comment (saiya) [reply jvm/apache-httpclient/README.md:20]: Okay, let's add @JvmStatic to help Java users.

By the way, this issue is same for `httpclient 4.2-`, right?
Do not hesitate to modify existing doc/code if some change needed.

File: jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt (hunk 3 above)
Comment (saiya) [reply jvm/apache-httpclient/src/main/kotlin/com/m3/tracing/apache/httpclient/ApacheHttpRequestInfo.kt]: In Kotlin, no need to write `.toString()` explicitly in the string interpolation.

```kotlin
"${foo.toString()}" == "${foo}"
```

File: jvm/apache-httpclient/README.md (hunk 1 above)
Comment (sekikatsu36) [reply jvm/apache-httpclient/README.md:20]: I overlooked the correction of the usage in `httpclient 4.2-`.
As same as `4.2+`, In this version we need to use `getInstance()` too, so I'd modified the description.

Comment (saiya) [reply jvm/apache-httpclient/README.md:20]: How about adding `@JvmStatic` in our library (from next release) ?

With `@JvmStatic`, no need to call `get` method even for Java users.

Comment (sekikatsu36) [reply jvm/apache-httpclient/README.md:20]: You mean `@JvmField`?
Currently the field has been annotated `@JvmStatic` already, and in this situation not enable to access as `.INSTANCE` as far as I tried.
(I intended to tell that in previous comments, but my comment to convey `@JvmField` might be bad. Sorry)
> To access INSTANCE, the field seems to be annotated @jvmField not @jvmStatic

Anyway, I agree with you. `.INSTANCE` is better than `.getInstance()` for Java users.
so I will change `@JvmStatic` to `@JvmField` in this PR.

Comment (saiya) [reply jvm/apache-httpclient/README.md:20]: Agreed.

Sorry for my confusion, I agree to use `@JvmStatic` (allow `.INSTANCE` static access) in this case.

**Comment by saiya** [comment]
LGTM!

**Pull request merged**